arangod/VocBase/LogicalDataSource.h

```diff
@@ -0,0 +1,114 @@
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Andrey Abramov
+/// @author Vasiliy Nabatchikov
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGOD__VOCBASE__LOGICAL_DATA_SOURCE_H
+#define ARANGOD__VOCBASE__LOGICAL_DATA_SOURCE_H 1
+
+#include "voc-types.h"
+#include "Basics/Result.h"
+
+struct TRI_vocbase_t; // forward declaration
+
+namespace arangodb {
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief a common ancestor to all database objects proving access to documents
+///        e.g. LogicalCollection / LoigcalView
+////////////////////////////////////////////////////////////////////////////////
+class LogicalDataSource {
+ public:
+
+  //////////////////////////////////////////////////////////////////////////////
+  /// @brief singleton identifying the underlying implementation type
+  ///        each implementation should have its own static instance
+  ///        once a type is emplace(...)ed it cannot be removed
+  //////////////////////////////////////////////////////////////////////////////
+  class Type final {
+   public:
+    Type(Type&& other) noexcept = default;
+    bool operator==(Type const& other) const noexcept { return this == &other; }
+    bool operator!=(Type const& other) const noexcept { return this != &other; }
+    static Type const& emplace(std::string&& name);
+    inline std::string const& name() const noexcept { return *_name; }
+
+   private:
+    std::string const* _name; // type name for e.g. log messages
+
+    Type() = default;
+    Type(Type const&) = delete;
+    Type& operator=(Type const&) = delete;
+    Type& operator=(Type&&) noexcept = delete;
+  };
+
+  LogicalDataSource(
+      Type const& type,
+      TRI_vocbase_t* vocbase,
+      TRI_voc_cid_t id,
+      TRI_voc_cid_t planId,
+      std::string&& name,
+      bool deleted
+  ) noexcept
+    : _name(std::move(name)),
+      _type(type),
+      _vocbase(vocbase),
+      _id(id),
+      _planId(planId),
+      _deleted(deleted) {
+  }
+  LogicalDataSource(LogicalDataSource const& other)
+    : _name(other._name),
+      _type(other._type),
+      _vocbase(other._vocbase),
+      _id(other._id),
+      _planId(other._planId),
+      _deleted(other._deleted) {
+  }
+
+  virtual ~LogicalDataSource() {}
+
+  inline bool deleted() const noexcept { return _deleted; }
+  virtual void drop() = 0;
+  inline TRI_voc_cid_t id() const { return _id; }
+  inline std::string const& name() const { return _name; }
+  inline TRI_voc_cid_t planId() const noexcept { return _planId; }
+  virtual Result rename(std::string&& newName, bool doSync) = 0;
+  inline Type const& type() const noexcept { return _type; }
+  inline TRI_vocbase_t* vocbase() const { return _vocbase; }
+
```

Comment:
The use of `noexcept` here seems rather inconsistent. For instance, why is `planId()` `noexcept`, but `id()` is not? We haven't come to a consensus as a team yet that we want to use `noexcept` on a regular basis, so I'd lean toward removing the specifiers here, at least for the time being.

Reply:
I'll address the noexcept consistency issue in one of the subsequent pull-requests for this task, there was divergence purely due to copy-paste  reasons, thank you for catching it